File: skills/docforge/_shared/runtime/manifest/js/hash_evidence.js
#!/usr/bin/env node
"use strict";
/** Stamp git_blob / git_blob_normalized / range_blob for one provenance source.
 *
 * The writing agent uses this to hash a cited file (and optionally a specific
 * line range) exactly the way `check_staleness.js` will later recompute it --
 * `git_blob` matches `git hash-object`; `git_blob_normalized` and `range_blob`
 * have no standard-tool equivalent, so both sides must share one implementation.
 */

const fs = require("fs");
const path = require("path");
const { fail } = require("../../common/js/_util.js");
const { normalizedBlobHash, rangeBlobHash, rawBlobHash } = require("../../common/js/evidence_hash.js");

function parseRange(value) {
  const match = /^([0-9]+)-([0-9]+)$/.exec(value);
  if (!match) throw new Error(`invalid --range: ${value} (expected <start>-<end>)`);
  const start = Number(match[1]);
  const end = Number(match[2]);
  if (start < 1 || end < start) throw new Error(`invalid --range: ${value} (expected 1 <= start <= end)`);
  return [start, end];
}

function hashEvidence(repo, relPath, span) {
  if (path.isAbsolute(relPath) || relPath.split(/[\\/]/).includes("..")) {
    throw new Error(`path escapes repo: ${relPath}`);
  }
  const target = path.resolve(repo, relPath);
  if (target !== repo && !target.startsWith(`${repo}${path.sep}`)) {
    throw new Error(`path escapes repo: ${relPath}`);
  }
  if (!fs.existsSync(target) || !fs.statSync(target).isFile()) {
    throw new Error(`file not found: ${relPath}`);
  }
  const content = fs.readFileSync(target);
  const result = { git_blob: rawBlobHash(content) };
  const normalized = normalizedBlobHash(content);
  if (normalized !== null) result.git_blob_normalized = normalized;
  if (span) {
    const [start, end] = span;
    const scoped = rangeBlobHash(content, start, end);
    if (scoped === null) {
      throw new Error(`cannot hash range ${start}-${end} of ${relPath} (out of bounds or not valid UTF-8 text)`);
    }
    result.evidence_range = { start: String(start), end: String(end) };
    result.range_blob = scoped;
  }
  return result;
}

function parseArgs(argv) {
  const result = {};
  const allowed = new Set(["repo", "path", "range", "json"]);
  for (let i = 0; i < argv.length; i++) {
    const token = argv[i];
    if (token === "-h" || token === "--help") return { help: true };
    if (!token.startsWith("--")) throw new Error(`unexpected argument: ${token}`);
    const raw = token.slice(2);
    if (!allowed.has(raw)) throw new Error(`unknown option: ${token}`);
    if (raw === "json") result[raw] = true;
    else {
      if (i + 1 >= argv.length || argv[i + 1].startsWith("--")) throw new Error(`option requires a value: ${token}`);
      result[raw] = argv[++i];
    }
  }
  return result;
}

function usage() {
  console.log("usage: hash_evidence.js --repo <repo> --path <repo-relative-path> [--range <start>-<end>] [--json]");
}

function main() {
  try {
    const args = parseArgs(process.argv.slice(2));
    if (args.help) { usage(); return 0; }
    if (!args.repo) throw new Error("--repo is required");
    if (!args.path) throw new Error("--path is required");
    const repo = path.resolve(args.repo);
    if (!fs.existsSync(repo) || !fs.statSync(repo).isDirectory()) {
      return fail(`not a directory: ${args.repo}`, 2);
    }
    let span = null;
    let result;
    try {
      span = args.range ? parseRange(args.range) : null;
      result = hashEvidence(repo, args.path, span);
    } catch (error) {
      return fail(error.message, 2);
    }
    if (args.json) {
      console.log(JSON.stringify(result, null, 2));
    } else {
      console.log(`git_blob: ${result.git_blob}`);
      if ("git_blob_normalized" in result) console.log(`git_blob_normalized: ${result.git_blob_normalized}`);
      if ("evidence_range" in result) {
        console.log(`evidence_range: ${result.evidence_range.start}-${result.evidence_range.end}`);
        console.log(`range_blob: ${result.range_blob}`);
      }
    }
    return 0;
  } catch (error) {
    usage();
    return fail(error.message, 2);
  }
}
process.exit(main());

File: skills/docforge/_shared/runtime/manifest/python/hash_evidence.py
#!/usr/bin/env python3
"""Stamp git_blob / git_blob_normalized / range_blob for one provenance source.

The writing agent uses this to hash a cited file (and optionally a specific
line range) exactly the way `check_staleness.py` will later recompute it --
`git_blob` matches `git hash-object`; `git_blob_normalized` and `range_blob`
have no standard-tool equivalent, so both sides must share one implementation.
"""

from __future__ import annotations

import argparse
import json
from pathlib import Path

from runtime.common.python._util import fail
from runtime.common.python.evidence_hash import (
    normalized_blob_hash,
    range_blob_hash,
    raw_blob_hash,
)


def parse_range(value: str) -> tuple[int, int]:
    start_text, sep, end_text = value.partition("-")
    if not sep or not start_text.isdigit() or not end_text.isdigit():
        raise ValueError(f"invalid --range: {value!r} (expected <start>-<end>)")
    start, end = int(start_text), int(end_text)
    if start < 1 or end < start:
        raise ValueError(f"invalid --range: {value!r} (expected 1 <= start <= end)")
    return start, end


def hash_evidence(repo: Path, rel_path: str, span: tuple[int, int] | None) -> dict:
    if Path(rel_path).is_absolute() or ".." in Path(rel_path).parts:
        raise ValueError(f"path escapes repo: {rel_path}")
    target = (repo / rel_path).resolve()
    if repo.resolve() not in (target, *target.parents):
        raise ValueError(f"path escapes repo: {rel_path}")
    if not target.is_file():
        raise ValueError(f"file not found: {rel_path}")
    content = target.read_bytes()
    result: dict = {"git_blob": raw_blob_hash(content)}
    normalized = normalized_blob_hash(content)
    if normalized is not None:
        result["git_blob_normalized"] = normalized
    if span is not None:
        start, end = span
        scoped = range_blob_hash(content, start, end)
        if scoped is None:
            raise ValueError(
                f"cannot hash range {start}-{end} of {rel_path} "
                "(out of bounds or not valid UTF-8 text)"
            )
        result["evidence_range"] = {"start": str(start), "end": str(end)}
        result["range_blob"] = scoped
    return result


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--repo", required=True, type=Path)
    parser.add_argument("--path", required=True, help="Repo-relative path to hash")
    parser.add_argument("--range", help="1-indexed inclusive line span, e.g. 10-20")
    parser.add_argument("--json", action="store_true")
    args = parser.parse_args()
    if not args.repo.is_dir():
        return fail(f"not a directory: {args.repo}", 2)
    try:
        span = parse_range(args.range) if args.range else None
        result = hash_evidence(args.repo.resolve(), args.path, span)
    except ValueError as exc:
        return fail(str(exc), 2)
    if args.json:
        print(json.dumps(result, indent=2, ensure_ascii=False))
    else:
        print(f"git_blob: {result['git_blob']}")
        if "git_blob_normalized" in result:
            print(f"git_blob_normalized: {result['git_blob_normalized']}")
        if "evidence_range" in result:
            print(f"evidence_range: {result['evidence_range']['start']}-{result['evidence_range']['end']}")
            print(f"range_blob: {result['range_blob']}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
